Pass system version to all hooks

obbardc opened the pull request:
Pass installed system version to bundle install-check hook script.

Fixes: #531

(This has been collecting dust downstream for a while now; time to get it upstreamed)

obbardc commented on src/install.c:
```diff
@@ -763,17 +763,28 @@ static gboolean run_bundle_hook(RaucManifest *manifest, gchar* bundledir, const
 
 	g_message("Running bundle hook %s", hook_cmd);
 
+	/* Pass the version of the currently running slot to the hook script */
```
This (4 year old!) logic is a bit awful, but it seems to work.

obbardc commented on src/install.c:
```diff
@@ -763,17 +763,29 @@ static gboolean run_bundle_hook(RaucManifest *manifest, gchar* bundledir, const
+	/* Pass the version of the currently running slot to the hook script */
+	s = find_config_slot_by_device(r_context()->config, r_context()->bootslot);
+	if (s) {
+		load_slot_status(s);
```
This is the wrong API, from an older version of RAUC. I will fix it.

jluebbe commented on src/install.c:
```diff
@@ -763,17 +763,28 @@ static gboolean run_bundle_hook(RaucManifest *manifest, gchar* bundledir, const
 
 	g_subprocess_launcher_setenv(launcher, "RAUC_SYSTEM_COMPATIBLE", r_context()->config->system_compatible, TRUE);
 	g_subprocess_launcher_setenv(launcher, "RAUC_SYSTEM_VARIANT", r_context()->config->system_variant ?: "", TRUE);
+	g_subprocess_launcher_setenv(launcher, "RAUC_SYSTEM_VERSION", system_version ?: "", TRUE);
```
By now, that part was factored out to `add_system_environment`, which you could use in `run_bundle_hook`.

jluebbe commented:
As mentioned in https://github.com/rauc/rauc/issues/531#issuecomment-1640254105, `RAUC_SYSTEM_VERSION` would be a misleading name. Each slot remembers the bundle version which was installed to it last, so there are potentially multiple versions (one per slot).

If I understood your request correctly, you need the version of the currently booted slot. With the info suggested in the linked comment, you could do something like this:
```
eval CURRENT_VERSION=\$RAUC_SLOT_BUNDLE_VERSION_${RAUC_BOOTED_SLOT_IDX}
```

Alternatively I'd also be fine with exporting a `RAUC_BOOT_SLOT_BUNDLE_VERSION` with that value.

obbardc committed:
src/main: show system version in output of rauc status

Signed-off-by: Christopher Obbard <christopher.obbard@linaro.org>

obbardc committed:
hooks: pass RAUC_SYSTEM_VERSION to all hooks

Export RAUC_SYSTEM_VERSION to install, bundle and slot hook
environments so hooks can access the system version returned by the
system-info handler.

Document the new environment variable in the reference manual for all
relevant hook types.

Fixes: #531
Signed-off-by: Christopher Obbard <christopher.obbard@linaro.org>

obbardc commented:
@jluebbe I changed approach here. I think it makes sense to expose this variable to hooks rather than the installed slot version, but let me know if I am missing something.

obbardc committed:
service: expose system version on D-Bus

Add a read-only SystemVersion property to the
de.pengutronix.rauc.Installer D-Bus interface and populate it from the
system version obtained from the system-info handler.

This makes the version returned by the system-info handler available to
D-Bus clients alongside the existing Compatible, Variant and BootSlot
properties.

Signed-off-by: Christopher Obbard <christopher.obbard@linaro.org>

obbardc commented on src/service.c:
```diff
@@ -612,6 +612,7 @@ static void r_on_bus_acquired(GDBusConnection *connection,
 
 	r_installer_set_compatible(r_installer, r_context()->config->system_compatible);
 	r_installer_set_variant(r_installer, r_context()->config->system_variant);
+	r_installer_set_system_version(r_installer, r_context()->system_version);
```
Not sure if `system_version` will be set at this point. Need to test it.

obbardc replied:
well, it does seems to work when calling `rauc status` !

obbardc commented:
@jluebbe can you take a look, please? This solves my usecase of getting the current system version in various hooks.

ejoerns commented:
In contrast to the original approach, the rework (using the 'system version' rather than the 'bundle version')  looks plausible to me.

A small difference to adjacent properties like the 'compatible' or the 'variant' still is that the system version refers to the *currently booted* system, not the device in general. So it might change when updating/rebooting.

But since the variable isn't named 'booted system version' in other context either, it might be fine to expose it as 'system version'.

obbardc commented:
> A small difference to adjacent properties like the 'compatible' or the 'variant' still is that the system version refers to the _currently booted_ system, not the device in general. So it might change when updating/rebooting.
> 
> But since the variable isn't named 'booted system version' in other context either, it might be fine to expose it as 'system version'.

Would calling this variable something like `current system version` or `booted system version` across the codebase be more descriptive ?
